File: Python/133. Clone Graph.py
"""
Given a reference of a node in a connected undirected graph.
Return a deep copy (clone) of the graph.
Each node in the graph contains a value (int) and a list (List[Node]) of its neighbors.
"""
# 133. Clone Graph
from typing import List
class Node:
    def __init__(self, val: int, neighbors: List['Node'] = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
        
from typing import Optional
class Solution:
    def cloneGraph(self, node: Optional['Node']) -> Optional['Node']:
        if not node:
            return None
        
        old_to_new = {}

        def dfs(n):
            if n in old_to_new:
                return old_to_new[n]

            copy = Node(n.val)
            old_to_new[n] = copy
            
            copy.neighbors = [dfs(neighbor) for neighbor in n.neighbors]
            return copy
        
        return dfs(node)
        
        
# Test Code
def printGraph(node: 'Node'):
    visited = set()
    def dfs(n):
        if n in visited:
            return
        visited.add(n)
        print(f"Node {n.val}: {[neighbor.val for neighbor in n.neighbors]}")
        for neighbor in n.neighbors:
            dfs(neighbor)
    dfs(node)


n1, n2, n3, n4 = Node(1), Node(2), Node(3), Node(4)
n1.neighbors = [n2, n3]
n2.neighbors = [n1, n3, n4]
n3.neighbors = [n1, n2]
n4.neighbors = [n2]

# ✅ Clone the Graph
solution = Solution()
cloned_graph = solution.cloneGraph(n1)

# ✅ Print the Original and Cloned Graphs
print("Original Graph:")
printGraph(n1)

print("\nCloned Graph:")
printGraph(cloned_graph)

File: PHP/133. Clone Graph.php
<?php
/*
Given a reference of a node in a connected undirected graph.
Return a deep copy (clone) of the graph.
Each node in the graph contains a value (int) and a list (List[Node]) of its neighbors.
*/
function printGraph($node) {
    if ($node === null) return "Graph is empty\n";

    $queue = [$node];
    $visited = [];

    while (!empty($queue)) {
        $curr = array_shift($queue);
        if (isset($visited[$curr->val])) continue;
        $visited[$curr->val] = true;

        $neighborVals = array_map(fn($n) => $n->val, $curr->neighbors);
        echo "Node {$curr->val} -> [" . implode(", ", $neighborVals) . "]\n";

        foreach ($curr->neighbors as $neighbor) {
            if (!isset($visited[$neighbor->val])) {
                $queue[] = $neighbor;
            }
        }
    }
}
    
class Node {
    public $val;
    public $neighbors;

    function __construct($val = 0, $neighbors = []) {
        $this->val = $val;
        $this->neighbors = $neighbors;
    }
}
    
class Solution {

    /**
     * 133. Clone Graph
     * @param Node $node
     * @return Node
     */
    public function cloneGraph(Node $node) {
        if ($node === null) { return null; }
        $visited = [];
        return $this->dfsClone($node, $visited);        
    }

    private function dfsClone(Node $node, array &$visited) {
        if (isset($visited[$node->val])) {
            return $visited[$node->val]; // Return already copied node
        }

        // Clone the current node
        $cloneNode = new Node($node->val);
        $visited[$node->val] = $cloneNode;

        // Recursively clone neighbors
        foreach ($node->neighbors as $neighbor) {
            $cloneNode->neighbors[] = $this->dfsClone($neighbor, $visited);
        }

        return $cloneNode;
    }
}

$node1 = new Node(1);
$node2 = new Node(2);
$node3 = new Node(3);
$node4 = new Node(4);

$node1->neighbors = [$node2, $node4];
$node2->neighbors = [$node1, $node3];
$node3->neighbors = [$node2, $node4];
$node4->neighbors = [$node1, $node3];

$c = new Solution;
$clonedGraph = $c->cloneGraph($node1);
echo "Original Graph:\n";
printGraph($node1);
echo "\nCloned Graph:\n";
printGraph($clonedGraph);
?>